cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/SparkContextCache.scala

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.app.runtime.spark
+
+import java.util.concurrent.ConcurrentLinkedQueue
+import javax.annotation.Nullable
+
+import org.apache.spark.SparkContext
+import org.apache.spark.scheduler._
+import org.apache.spark.streaming.StreamingContext
+
+import scala.collection.JavaConversions._
+
+/**
+  * A singleton cache for [[org.apache.spark.SparkContext]] and [[org.apache.spark.streaming.StreamingContext]]
+  * to provide universal access to those contexts created by the user Spark program within CDAP.
+  *
+  * With ClassLoader isolation, there will be one instance of this class per one Spark program execution.
+  */
+object SparkContextCache {
+
+  private var stopped = false
+  private var sparkContext: Option[SparkContext] = None
+  private var streamingContext: Option[StreamingContext] = None
+  private val sparkListeners = new ConcurrentLinkedQueue[SparkListener]()
+
+  /**
+    * Adds a [[org.apache.spark.scheduler.SparkListener]]. The given listener will be added to
+    * [[org.apache.spark.SparkContext]] when it becomes available.
+    */
+  def addSparkListener(listener: SparkListener): Unit = sparkListeners.add(listener)
+
+  /**
+    * Sets the [[org.apache.spark.SparkContext]] for the execution.
+    */
+  def setContext(context: SparkContext): Unit = {
+    this.synchronized {
+      if (stopped) {
+        context.stop()
+        throw new IllegalStateException("Spark program is already stopped")
+      }
+
+      if (sparkContext.isDefined) {
+        throw new IllegalStateException("SparkContext was already created")
+      }
+
+      sparkContext = Some(context)
+      context.addSparkListener(new DelegatingSparkListener)
+    }
+  }
+
+  /**
+    * Sets the [[org.apache.spark.streaming.StreamingContext]] that is currently in use.
+    */
+  def setContext(context: StreamingContext): Unit = {
+    if (stopped) {
+      context.stop(false)
+      throw new IllegalStateException("Spark program is already stopped")
+    }
+
+    // Spark doesn't allow multiple StreamingContext instances concurrently, hence we don't need to check in here
```

Comment:
Do you mean we shouldn't allow it? The current code only captures one `StreamingContext`, meaning if Spark really allows multiple streaming context concurrently, we won't be able to stop all of them. However, the user can still stop it themselves. 

We can probably modify the code to track all streaming context using a concurrent set, although it seems a bit overkill for now.